src/ripple/nodestore/impl/BatchWriter.cpp

```diff
@@ -39,7 +39,12 @@ BatchWriter::~BatchWriter ()
 void
 BatchWriter::store (std::shared_ptr<NodeObject> const& object)
 {
-    std::lock_guard<decltype(mWriteMutex)> sl (mWriteMutex);
+    std::unique_lock<decltype(mWriteMutex)> sl (mWriteMutex);
+
+    // If the batch has reached its limit, we wait
+    // until the batch writer is finished
+    while (mWriteSet.size() >= batchWriteLimitSize)
```

Comment:
While you are in the neighborhood... You may consider declaring and reserving `set` before the loop in `::writeBatch` and clearing it with each iteration. That would be an improvement with batches near the `batchWriteLimitSize` threshold.

Reply:
I'm pretty sure we do that. We have a reservation parameter that's 128 or 256 and we should be using it.

Reply:
We reserve 256 and possibly reallocate up to 65536 with each iteration. On heavy write loads, it would be nice to simply reuse.